interfaces/builtin/docker.go

```diff
@@ -0,0 +1,185 @@
+
+# Docker needs to be able to create and load the profile it applies to containers ("docker-default")
+# XXX We might be able to get rid of this if we generate and load docker-default ourselves and make docker not do it.
+/etc/apparmor.d/docker rw,
+/sbin/apparmor_parser ixr,
+/etc/apparmor*/** r,
+/etc/apparmor.d/cache/docker* w,
+/sys/kernel/security/apparmor/** rw,
+
+# We'll want to adjust this to support --security-opts...
+change_profile -> docker-default,
+signal (send) peer=docker-default,
+ptrace (read, trace) peer=docker-default,
+
+# Graph (storage) driver bits
+/dev/shm/aufs.xino rw,
+
+#cf bug 1502785
+/ r,
+`)
+
+var dockerConnectedPlugAppArmor = []byte(`
+# Description: Allow using Docker. Reserved because this gives
+#  privileged access to the service/host.
+# Usage: reserved
+
+# Obviously need to be able to talk to the daemon
+/{,var/}run/docker.sock rw,
+
+@{PROC}/sys/net/core/somaxconn r,
+`)
+
+var dockerPermanentSlotSecComp = []byte(`
+# The Docker daemon needs to be able to launch arbitrary processes within containers (whose syscall needs are unknown beforehand)
+@unrestricted
+`)
+
+var dockerConnectedPlugSecComp = []byte(`
+setsockopt
+bind
+`)
+
+type DockerInterface struct{}
+
+func (iface *DockerInterface) Name() string {
+	return "docker"
+}
+
+func (iface *DockerInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecurityDBus, interfaces.SecurityMount, interfaces.SecuritySecComp, interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		return dockerConnectedPlugAppArmor, nil
+	case interfaces.SecuritySecComp:
+		return dockerConnectedPlugSecComp, nil
+	case interfaces.SecurityDBus, interfaces.SecurityMount, interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		return dockerPermanentSlotAppArmor, nil
+	case interfaces.SecuritySecComp:
+		return dockerPermanentSlotSecComp, nil
+	case interfaces.SecurityDBus, interfaces.SecurityMount, interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecurityDBus, interfaces.SecurityMount, interfaces.SecuritySecComp, interfaces.SecurityUDev:
```

Comment:
Can you add a comment here: '// The docker socket is a named socket and therefore mediated by AppArmor file rules and we can't currently limit connecting clients by their security label'